linkerd/proxy/http/src/client.rs

```diff
@@ -128,13 +130,16 @@ where
             } => {
                 let exec =
                     tokio::executor::DefaultExecutor::current().instrument(info_span!("http1"));
+
+                let abs_form = target.should_overwrite_authority() || was_absolute_form;
+
                 let h1 = hyper::Client::builder()
                     .executor(exec)
                     .keep_alive(keep_alive)
-                    // hyper should never try to automatically set the Host
-                    // header, instead always just passing whatever we received.
-                    .set_host(false)
-                    .build(HyperConnect::new(connect, target, was_absolute_form));
+                    // hyper should should only try to automatically
+                    // set the host if the request was in absolute_form
+                    .set_host(abs_form)
+                    .build(HyperConnect::new(connect, target, abs_form));
```

Comment:
OK, Help me understand thsi change in behavior. We always created this with `false` before, but now it's not only conditioned on overwriting, but also on the `was_absolute_form` value? I'm still getting caught up on this.

Reply:
1. Why is it always appropriate to do this when we are overwriting?
2. It feels leaky that the http client has to know about authority overwriting at all. I'd expect the target's `HasSettings` value to be enough to configure whatever is needed by the client.

Reply:
yeah I was a bit confused by that myself, but let me explain my logic. First of all... the gateway itself at the moment routes based on the host header of a request. So really what we want to happen is for that host header value to be equal to the fqdn of the servie on the other side. So what happens in our case when we have proxy to proxy (the gateway is meshed as well): 

On the outbound side:
- We get the overwrite metadata
- In the overwrite_authority we blank the canonical_dst and host headers AND we set the authority of the request.
- We upgrade the request of the to H2 from H1 (because we can :). 
- When we build the client, we enable automatic host header derivation (based on authority) IF we have overwritten the authority. Now keep in mind that this is a setting that only applies to H1 clients. Because we have upgraded to H2 this is effectivelly a no-op in this scenario.
- The request goes on to the remote cluster gateway with the correct authority

On the inbound side: 
- We downgrade the requests to H1.
- Because this H1 request arrived in abs form on the inbound side, we do .set_host(true) for hyper
- Now hyper will derive the Host header based on the authority and that request will arrive at the local application (the gateway) with the Host header set to the FQDN name of the service that need sto receive it at the end. 

I made the changes that you suggested so we pipe less dependencies into the client